### PR TITLE
[R4R]-{estimateGas}feature: EstimateGas performance optimization

### PR DESCRIPTION
**Pull request by boz14676**
EstimateGas performance optimization:

Reference：https://github.com/ethereum/go-ethereum/blob/master/eth/gasestimator/gasestimator.go#L52
1. Introduction of Optimistic Gas Limit
2. Finely Tuned Binary Search
3. Handling of Fee Caps
4. Rapid Processing of Special Cases
5. Joint Consideration of Account Balance and Gas Limits
6. Error Handling and Early Return

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 	"encoding/hex"
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/consensus"
```
Comment (Tri-stone) [inline internal/ethapi/api.go:24]: go import format

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -52,8 +52,11 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 )
 
-var (
-	gasBuffer = uint64(120)
+// estimateGasErrorRatio is the amount of overestimation eth_estimateGas is
+// allowed to produce in order to speed up calculations.
+const (
+	estimateGasErrorRatio = 0.015
+	gasBuffer             = uint64(120 / 100)
```
Comment (Tri-stone) [inline internal/ethapi/api.go:59]: `uint64(120 / 100)` will be `1`, it can not enlarge the estimated gas

File: internal/ethapi/api_test.go [hunk 3]
```diff
@@ -17,9 +17,28 @@
 package ethapi
 
 import (
+	"context"
+	"crypto/ecdsa"
 	"encoding/json"
+	"errors"
+	"github.com/ethereum/go-ethereum"
```
Comment (Tri-stone) [inline internal/ethapi/api_test.go:24]: go format

File: internal/ethapi/api.go (hunk 1 above)
Comment (boz14676) [reply internal/ethapi/api.go]: resolved

File: .gitignore [hunk 4]
```diff
@@ -49,3 +49,4 @@ profile.cov
 /dashboard/assets/package-lock.json
 
 **/yarn-error.log
+node_modules
```
[the remark this answers is not in the document]
Comment (boz14676) [reply .gitignore]: resolved

File: internal/ethapi/api.go (hunk 1 above)
Comment (boz14676) [reply internal/ethapi/api.go]: resolved

**Review by byteflyfunny — APPROVED** [review]

**Review by Tri-stone — APPROVED** [review]

**Pull request merged**